libvast/src/system/index.cpp

```diff
@@ -214,212 +193,346 @@ caf::actor index_state::next_worker() {
-    else if (auto ptr = find_unpersisted(partition_id); ptr != nullptr)
-      part = ptr;
-    else
-      part = lru_partitions.get_or_add(partition_id).get();
-    auto eval = part->eval(lookup.expr);
-    if (eval.empty()) {
-      VAST_DEBUG(self, "identified partition", partition_id,
-                 "as candidate in the meta index, but it didn't produce an "
-                 "evaluation map");
-      return;
-    }
-    result.emplace(partition_id, std::move(eval));
+    // of our unpersisted ones. Only then can we dispatch to our LRU cache.
+    caf::actor part;
+    if (active_partition.actor != nullptr
+        && active_partition.id == partition_id)
+      part = active_partition.actor;
+    else if (auto it = unpersisted.find(partition_id); it != unpersisted.end())
+      part = it->second;
+    else if (auto it = persisted_partitions.find(partition_id);
+             it != persisted_partitions.end())
+      part = inmem_partitions.get_or_load(partition_id);
+    if (!part)
+      VAST_ERROR("Could not load partition", partition_id,
+                 "that was part of a query");
+    return part;
   };
   // Loop over the candidate set until we either successfully scheduled
   // num_partitions partitions or run out of candidates.
-  {
-    auto i = lookup.partitions.begin();
-    auto last = lookup.partitions.end();
-    for (; i != last && result.size() < num_partitions; ++i)
-      spin_up(*i);
-    lookup.partitions.erase(lookup.partitions.begin(), i);
+  auto it = lookup.partitions.begin();
+  auto last = lookup.partitions.end();
+  while (it != last && result.size() < num_partitions) {
+    auto partition_id = *it++;
+    if (auto partition_actor = spin_up(partition_id))
+      result.push_back(std::make_pair(partition_id, partition_actor));
   }
+  lookup.partitions.erase(lookup.partitions.begin(), it);
+  VAST_DEBUG(self, "launched", result.size(),
+             "await handlers to fill the pending query map");
   return result;
 }
 
-query_map
-index_state::launch_evaluators(pending_query_map pqm, expression expr) {
-  query_map result;
-  for (auto& [id, eval] : pqm) {
-    std::vector<caf::actor> xs{self->spawn(evaluator, expr, std::move(eval))};
-    result.emplace(id, std::move(xs));
+/// Sends an expression to all partition actors passed into this function,
+/// and collects the resulting evaluators.
+/// @param c Continuation that takes a single argument of type
+/// `caf::expected<query_map>`. The continuation will be called in the
+/// actor context of `self`.
+//
+// TODO: At some point we should add some more template magic on top of
+// this and turn it into a generic functions that maps
+//
+//   (map from U to A, request param pack R, result handler with param X) ->
+//   expected<map from U to X>
+template <typename Continuation>
+void await_evaluators_then(
+  caf::stateful_actor<index_state>* self, const expression& expr,
+  const std::vector<std::pair<vast::uuid, caf::actor>>& actors,
+  Continuation c) {
```

Comment:
```suggestion
  Continuation continuation) {
```

Sticks out more that way down below (since `c` is not an idiomatic function name).

Reply:
Called it `then` instead, which feels more idiomatic.